### ISSUE TITLE
DbCommand failure message doesn't include exception details

### ISSUE THREAD

**Issue by clement911** [dotnet/efcore#24530]
Version: EF Core 5.0

We see in our logs many command error messages.

It says "Failed executing DbCommand ..."

https://github.com/dotnet/efcore/blob/26c0d2a8f3eab7d92ae9b7bb9dc7aa31a336bd90/src/EFCore.Relational/Properties/RelationalStrings.resx#L449-L452

We want to know what the exception is so that we can troubleshoot the problem.
However, it seems that EF core doesn't include the exception in the logged message.

https://github.com/dotnet/efcore/blob/809fe7a219c99fbe4f3576ef42bd084d4a6ce056/src/EFCore.Relational/Diagnostics/RelationalLoggerExtensions.cs#L1124-L1139

Was this an oversight or is this by design?
How can we see the exception?

**Comment by ajcvickers** [comment]
@clement911 This is by-design, since EF doesn't catch these exceptions (except see #15269) but rather lets the exception propagate, or sometimes wraps it in an outer exception and throws that. Therefore at some point either the application will crash, at which point the full exception message and stack trace should be generated, or something else is swallowing the exception without logging the details.

**Comment by clement911** [comment]
@ajcvickers i dont believe it ever bubbles up in our case. Is it possible it gets handles by the retry strategy?
In other words, does the retry strategy log  errors before retrying?
That would make sense but we'd still like to know why there was a failure.

**Comment by ajcvickers** [comment]
Yes, if you are using [connection resiliency](https://docs.microsoft.com/en-us/ef/core/miscellaneous/connection-resiliency), then this may be an error than then gets retried. #15269 is tracking making it possible to determine this.

**Comment by clement911** [comment]
Thank you @ajcvickers  but I believe the two issues are different.

#15269 is about determining whether an error was caused by a transient/retried error or not.

However, what I'm asking for is the exception details of transient errors.
The fact that it is transient doesn't mean that I don't want to know the exact error that occurred.
For example, if the transient error was due to a deadlock victim retry, it could indicate that the database is struggling with the current scale of concurrent requests. It is definitely something I would want to know early, rather than waiting for such times when even retrying N times doesn't help and the error does bubble up.

So, why not include the exception in transient errors?

**Comment by ajcvickers** [comment]
We discussed this and came to the following conclusions:
* Exceptions that we understand but do not handle will continue to be logged at `Error` level and will include the exception/stack trace. This may result in duplicates in the log, since the component that ultimate handles the error should also log it. However, it still seems better to be clear than an error has happened, and not rely on some other component to do that.
* We will continue to not log expected exceptions, such as when checking database existence for a database that does not exist.
* For transient errors that are retried we will:
  * Log a message at the `Info` level.
  * Include details of the transient failure (possibly the full exception and stack traces)
  * Include other relevant information, such as the retry count
 
The transient error changes will be tracked by #15269

Note that default logging levels can always be changed by the application.

**Comment by clement911** [comment]
Glad to see this is getting some attention.
Please do include the full exception details, otherwise that doesn't help what this issue is about.
Will there be an extension point to change the log level of all transient errors specifically (we would like Warning and not Info).

**Comment by clement911** [comment]
We're still seeing a lot of errors in the logs "Failed executing DbCommand" for transient errors but the lack of exception details makes it impossible to know why there was a failure.

**Comment by clement911** [comment]
We are still struggling with this issue. Any chance it could be added to the roadmap?

**Comment by clement911** [comment]
@ajcvickers , is it possible to add this to the EF7 roadmap?

**Comment by vslee** [comment]
@clement911, depending on how reproducible your error is, a workaround might be to run debug mode and have Visual Studio break on all exceptions. Then you can see the exception details are, even if it is caught.

**Comment by clement911** [comment]
@vslee we need this in production to see errors in our logs. This is not a development time concern

**Comment by atrimbl** [comment]
We implemented cancellation tokens to a large portion of our requests today and were immediately flooded with these errors in our logs. The cancellation tokens work as intended and kill the db connection, but it does result in this generic error.

**Comment by SimonCropp** [comment]
+1 one on including the exception details+stack trace. we get this intermittently in production logs. but there is no subsequent bubbling does not seem to happen. so it is likely some buggy code that is doing the command and catching. but we would like to know where it is coming from

**Comment by roji** [comment]
@atrimbl are you using 7.0? 7.0 distinguishes cancellation from failure ([see docs](https://learn.microsoft.com/en-us/ef/core/what-is-new/ef-core-7.0/whatsnew#distinguish-cancellation-from-failure-in-logs)).

**Comment by clement911** [comment]
@roji this issue is not about distinguishing cancellations. We want to get exception details of transient failures.

**Comment by roji** [comment]
@clement911 I know, I was specifically responding to @atrimbl complaint about cancellations above.

**Comment by roji** [comment]
Everyone, be sure to upvote this issue if you want it - that informs our prioritization process in a significant way.

**Comment by clement911** [comment]
@roji sorry about that, I thought you were responding to the next message.

**Comment by clement911** [comment]
@roji  @ajcvickers is there any chance this could be added to the roadmap?

All transient errors are logged but with no details of the actual exception.
We see transient errors in production and don't know why they occur.

We don't know why. Is it a deadlock victim? A timeout? Something else? We want to know.

**Comment by ajcvickers** [comment]
@clement911 This issue currently has 9 votes (👍). While this isn't the only factor in deciding whether we work on an issue, it does mean that currently around 170 other issues have more votes.

**Comment by clement911** [comment]
I understand but I figured this was probably a relatively small task. Since the failure is already logged, isn't it a simple task to add the exception with it?

**Comment by clement911** [comment]
I'm not sure sure where this is sitting in the priority. I see it got a few more votes.
I just wanted to mention that, as we scale our app, this is becoming an increasing issue. We hope will be addressed in the near future.

**Comment by Xriuk** [comment]
+1 on this, sometimes we see this error popup in logs, where the command is a simple SELECT which works the other 90% of the times, and when it appears it's not clear if it is a resiliency problem or what.
The original exception is needed to be logged as well.

**Comment by bharatmane** [comment]
I am also struggling with this issue especially while inserting. and not able to find any way to understand what's happening

**Comment by FraktalInverTer** [comment]
We updated the EF dlls from 7.0.1 to 7.0.11 a few days ago and starting getting these errors in log. No exception, no stack trace, yet logged at error level. I try catching the exceptions where I suspect it might be happening but I don't find anything bubbling up. I try attaching to SaveChangesFailed event but never get there. Is there any chance that you still have transient errors that are marked as errors instead of Info? How can I figure out what is actually happening? Program seems to be running just fine after error, so I assume it is handled by the EF somehow through retries, but if I don't know what is happening I can't try to fix any root cause, and it is spamming the log.
Any help appreciated.

**Comment by martinmine** [comment]
We are seeing a lot of these on SELECT-statements in production, but it is unclear to us why they are occurring due to lack of information in the error message that we see in the logs. From the timeout parameters we can see that it is not a timeout that takes place. Since we have enabled EnableRetryOnFailure, the queries gets re-tried and no requests actually fail. However, from an operational perspective this can potentially trigger alerts as we get error events in our logs. Is there any way to approach this in order to figure out what is going on?

**Comment by clement911** [comment]
We continue to suffer from this.
With now 40 votes, is it any closer to being prioritized?

**Comment by ajcvickers** [comment]
> With now 40 votes, is it any closer to being prioritized?

45 votes gets you to about position 57 in the list of most voted for issues. This in-of-itself is not enough to prioritize it as a highly-requested feature.

**Comment by stevendarby** [comment]
@clement911 below the code highlighted in your second link, it seems like an interceptor is called and the exception is passed to it. Could you create a command failed interceptor that logs the extra details you need?

**Comment by clement911** [comment]
@stevendarby I'm not sure why you mean. Where do you see an call to an interceptor?

**Comment by stevendarby** [comment]
@clement911 https://github.com/dotnet/efcore/blob/809fe7a219c99fbe4f3576ef42bd084d4a6ce056/src/EFCore.Relational/Diagnostics/RelationalLoggerExtensions.cs#L1162

I’m thinking if you register an ‘IDbCommandInterceptor’ that implements `CommandFailedAsync`, this could be somewhere you log the exception. More about interceptors here: https://learn.microsoft.com/en-us/ef/core/logging-events-diagnostics/interceptors

**Comment by valentasm1** [comment]
> @clement911
> 
> https://github.com/dotnet/efcore/blob/809fe7a219c99fbe4f3576ef42bd084d4a6ce056/src/EFCore.Relational/Diagnostics/RelationalLoggerExtensions.cs#L1162
> 
> I’m thinking if you register an `IDbCommandInterceptor` that implements `CommandFailedAsync`, this could be somewhere you log the exception. More about interceptors here: https://learn.microsoft.com/en-us/ef/core/logging-events-diagnostics/interceptors

Can confirm. This one helped me. Just not sure CommandFailedAsync or CommandFailed version. Also i think you should log it and not rethrow.

**Comment by XaspMe** [comment]
Still waiting for a fix, I came across this on my project and was surprised that it has been hanging there for three years.

**Comment by phil-willowtree** [comment]
Still seeing this issue. Any updates on a fix?

**Comment by roji** [comment]
This issue is in the Backlog milestone. This means that it is not planned for the next release (EF Core 10.0). We will [re-assess the backlog](https://docs.microsoft.com/ef/core/what-is-new/release-planning) following the this release and consider this item at that time. However, keep in mind that there are many other high priority features with which it will be competing for resources. Make sure to vote (👍) for this issue if it is important to you.
